crypto/rsa/rsa_sign.c

```diff
@@ -16,60 +16,189 @@
+        *len = sizeof(id_##name##_oid); \
+        return id_##name##_oid
+#define MD_CASE_NN(name) \
+    case NID_##name: \
+        *len = sizeof(id_##name##_oid_nn); \
+        return id_##name##_oid_nn
+
+#ifndef FIPS_MODE
+MD_DATA(md2, 0x02, 0x10);
+MD_DATA(md5, 0x05, 0x10);
+
+/* SHA-1 (1 3 14 3 2 26) */
+static const unsigned char id_sha1_oid[] = {
+    0x30, 0x21, 0x30, 0x09, 0x06, 0x05, 0x2b, 0x0e, 0x03, 0x02, 0x1a, 0x05,
+    0x00, 0x04, 0x14
+};
+
+/* SHA-1 (1 3 14 3 2 26), no NULL parameters */
+static const unsigned char id_sha1_oid_nn[] = {
+    0x30, 0x1f, 0x30, 0x07, 0x06, 0x05, 0x2b, 0x0e, 0x03, 0x02, 0x1a, 0x04,
+    0x14
+};
+
+/* MDC-2 (2 5 8 3 101) */
+static const unsigned char id_mdc2_oid[] = {
+    0x30, 0x1c, 0x30, 0x08, 0x06, 0x04, 0x55, 0x08, 0x03, 0x65, 0x05, 0x00,
+    0x04, 0x10
+};
+#endif
+
+SHA_DATA(sha256, 0x01, 0x20);
+SHA_DATA(sha384, 0x02, 0x30);
+SHA_DATA(sha512, 0x03, 0x40);
+SHA_DATA(sha224, 0x04, 0x1c);
+SHA_DATA(sha512_224, 0x05, 0x1c);
+SHA_DATA(sha512_256, 0x06, 0x20);
+SHA_DATA(sha3_224, 0x07, 0x1c);
+SHA_DATA(sha3_256, 0x08, 0x20);
+SHA_DATA(sha3_384, 0x09, 0x30);
+SHA_DATA(sha3_512, 0x0a, 0x40);
+
+static const unsigned char *digestinfo_encoding(int nid, size_t *len)
+{
+    switch (nid) {
+#ifndef FIPS_MODE
+    MD_CASE(mdc2);
+    MD_CASE(md2);
+    MD_CASE(md5);
+    MD_CASE(sha1);
+#endif
+    MD_CASE(sha224);
+    MD_CASE(sha256);
+    MD_CASE(sha384);
+    MD_CASE(sha512);
+    MD_CASE(sha512_224);
+    MD_CASE(sha512_256);
+    MD_CASE(sha3_224);
+    MD_CASE(sha3_256);
+    MD_CASE(sha3_384);
+    MD_CASE(sha3_512);
+    default:
+        return NULL;
+    }
+}
+
+/*
+ * Alternative DigestInfo encoding without parameters.
+ * The parameters may be not present (i.e. not NULL) when a SHA digest is used.
+ */
+static const unsigned char *digestinfo_encoding_nn(int nid, size_t *len)
+{
+    switch (nid) {
+#ifndef FIPS_MODE
+    MD_CASE_NN(sha1);
+#endif
+    MD_CASE_NN(sha224);
+    MD_CASE_NN(sha256);
+    MD_CASE_NN(sha384);
+    MD_CASE_NN(sha512);
+    MD_CASE_NN(sha512_224);
+    MD_CASE_NN(sha512_256);
+    MD_CASE_NN(sha3_224);
+    MD_CASE_NN(sha3_256);
+    MD_CASE_NN(sha3_384);
+    MD_CASE_NN(sha3_512);
+    default:
+        return NULL;
+    }
+}
+
 /* Size of an SSL signature: MD5+SHA1 */
 #define SSL_SIG_LENGTH  36
 
 /*
  * encode_pkcs1 encodes a DigestInfo prefix of hash |type| and digest |m|, as
  * described in EMSA-PKCS1-v1_5-ENCODE, RFC 3447 section 9.2 step 2. This
  * encodes the DigestInfo (T and tLen) but does not add the padding.
+ * 
+ * When verifying, decrypt_len indicates the size of the decrypted DigestInfo
+ * and is used to determine if the alternative encoding should be used for
+ * comparison.  When signing, decrypt_len is zero.
  *
  * On success, it returns one and sets |*out| to a newly allocated buffer
  * containing the result and |*out_len| to its length. The caller must free
  * |*out| with |OPENSSL_free|. Otherwise, it returns zero.
```

Comment:
Maybe don't use pipes around the function name, use () after it?

Reply:
Okay. That was already like that, but I will fix it anyway. :)